crates/matrix-sdk/src/config/request.rs

```diff
@@ -78,10 +106,22 @@ impl RequestConfig {
     }
 
     /// Create a new `RequestConfig` with default values, except the retry limit
-    /// which is set to 3.
+    /// which is set to `short_retry_limit`, or 3 by default.
     #[must_use]
-    pub fn short_retry() -> Self {
-        Self::default().retry_limit(3)
+    pub fn short_retry(&self) -> Self {
+        match self.short_retry_limit {
+            Limit::Unlimited => Self::default(),
+            Limit::Default => Self::default().retry_limit(3),
+            Limit::Limit(v) => Self::default().retry_limit(v),
+        }
+    }
```

Comment:
Well, to achieve the same behavior as previously, unless I've made a mistake somewhere, we can write
```rust
RequestConfig::new().short_retry()
```

Reply:
That being said, I'm not a big fan of what I did with the API. I'm sure that there is a better way to do things.

Reply:
My biggest problem is that `short_retry` doesn't convey a lot of information about what it actually does: Resetting all configurations except the “retry” configuration. And in the actual code, `short_retry` is a constructor, and now it's a statefull method that constructs a new `RequestConfig`.

Since it resets everything except the “retry” configuration, why not… hmm… renaming it `to_self_with_short_retry_only`, or `to_short_retry` (I like none of them). Using `into` would assume it will consume `RequestConfig`, which is wrong. Using `as` would assume it will cast it, which is wrong. Using `to` assumes it produces something new.
Another idea: `to_self_keep_short_retry`. I don't have a lot of ideas right now.

Reply:
So what we currently have is that we can pass in a `RequestConfig` to the `ClientBuilder`:
https://github.com/matrix-org/matrix-rust-sdk/blob/04d326eec176636c5039682132b92bd2affd7dec/crates/matrix-sdk/src/client/builder.rs#L171

The `RequestConfig` we pass in to the builder then gets stored as part of the `HttpClient`, if no `RequestConfig` is passed to the builder we use the default one:

https://github.com/matrix-org/matrix-rust-sdk/blob/main/crates/matrix-sdk/src/http_client.rs#L96=

Now we mainly use that `RequestConfig` stored in the `HttpClient`, unless we override it in the `Client` layer, like for the `/login` call:

https://github.com/matrix-org/matrix-rust-sdk/blob/04d326eec176636c5039682132b92bd2affd7dec/crates/matrix-sdk/src/client/login_builder.rs#L115

What I think makes sense here is to have a completely separate `RequestConfig` that will be used for the short retry cases.

This would mean that we need to do a couple of things:
1. Make the `HttpClient` stateless, so it doesn't hold on to the `RequestConfig` anymore, by moving the `RequestConfig` from the `HttpCLient` to the `Client`.
2. Add another field to the `Client` that holds the short retry version of the `RequestConfig`
3. Allow users to set the short retry version of the `RequestConfig` via the `ClientBuilder` as well.

This way users can define what the usual `RequestConfig` looks like an what the short retry `RequestConfig` looks like.

There were already talks about making `HttpClient` stateless, though that may have been motivated by https://github.com/matrix-org/matrix-rust-sdk/issues/805.

Reply:
ping @Yoric, do you plan to get back to this?